### ISSUE TITLE
InstructBLIP requiring eager attention specified to work properly

### ISSUE THREAD

**Issue by nhatkhtn** [huggingface/transformers#40952]
### System Info

The model `instructblip-vicuna-7b` gives wildly different results when specifying `attn_implementation='eager'` and not specifying anything. Furthermore, when specifying eager attention, the model seems to work correctly, while not specifying gives valid but noticeably worse results. For example, when captioning a COCO image:

- Running with specifying eager: "The image features a giraffe standing near a tree, with its reflection visible in the water. The giraffe is positioned in the center of the scene, with its head and neck stretched upwards towards the tree. There are several trees scattered throughout the scene, with one tree located on the left side of the image and another on the right side. The water body is located near the giraffe, with its reflection visible in the water. The overall scene appears to be a peaceful and serene setting, with the giraffe standing tall and majestic amidst the trees and water"
- Running without specifying eager: "a reflection of a tree in a lake with a blue sky and clouds"

System info:
- `transformers` version: 4.52.4
- Platform: Linux-5.14.0-503.40.1.el9_5.x86_64-x86_64-with-glibc2.34
- Python version: 3.12.11
- Huggingface_hub version: 0.33.0
- Safetensors version: 0.5.3
- Accelerate version: 1.7.0
- Accelerate config:    not found
- DeepSpeed version: not installed
- PyTorch version (GPU?): 2.7.1+cu128 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: Yes
- GPU type: NVIDIA B200

### Who can help?

@zucchini-nlp 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Run the following code, with and without the `attn_implementation="eager",` line
```
import requests
import torch
from PIL import Image
from transformers import InstructBlipProcessor, InstructBlipForConditionalGeneration

device = "cuda"

model = InstructBlipForConditionalGeneration.from_pretrained(
    "Salesforce/instructblip-vicuna-7b",
    attn_implementation="eager",
).to(device)
processor = InstructBlipProcessor.from_pretrained("Salesforce/instructblip-vicuna-7b")

url1 = "http://images.cocodataset.org/val2014/COCO_val2014_000000030925.jpg"
image1 = Image.open(requests.get(url1, stream=True).raw).convert("RGB")

prompt = "Please help me describe the image in detail."
inputs = processor(images=image1, text=prompt, return_tensors="pt", padding=True).to(device)

with torch.inference_mode():
    outputs = model.generate(
        **inputs,
        max_length=256,
    )
texts = processor.batch_decode(outputs, skip_special_tokens=True)
texts

```

### Expected behavior

Since InstructBLIP does not support any attention implementation other than eager, the output when specifying attn_implementation='eager' and when not specifying anything should be the same.

**Comment by zucchini-nlp** [comment]
InstructBlip actually supports all attention implementations and by default it will dispatch on SDPA which can cause minor differences in logits. The only part of instructBlip with no SDPA support is the qformers because it has non-standard position embeddings

Would be helpful, if you can also check the hidden states of vision and language model after one forward pass to calculate the difference. I will be able to dig into it a it later tomorrow

**Comment by zucchini-nlp** [comment]
I see you are using `transformers version: 4.52.4`. Can you update with `pip install -U transformers`? The issue was fixed apparently already in subsequent releases and with `4.56.2` I am getting this text without indicating attention implementation

```
['Please help me describe the image in detail. The image features a giraffe standing near a tree, with its reflection visible in the water. The giraffe is positioned in the center of the scene, with its head and neck stretched upwards towards the tree. There are several trees scattered throughout the scene, with one tree located on the left side of the image and another on the right side. The water body is located near the giraffe, with its reflection visible in the water. The overall scene appears to be a peaceful and serene setting, with the giraffe standing tall and majestic amidst the trees and water.']
```

**Comment by nhatkhtn** [comment]
Thank you for the fast response! Since setting attn_implementation to anything other than eager doesn't work, I assumed that the model has to have used eager, which is not the case.

Anyway, I can confirm that updating transformers to `4.56.1` fixes the problem for me. Regardless, it is helpful for the community to be aware of and thus avoid this issue.

**Comment by zucchini-nlp** [comment]
Great, then we can close the issue :)
